File: agroamigo-app/src/lib/cache.ts
/**
 * In-memory session cache for query results. Keyed by string, TTL-bounded,
 * invalidate-by-prefix. Mirror of packages/shared/src/lib/cache.ts — the
 * mobile app is not a workspace member, so we keep a local copy rather than
 * add metro monorepo plumbing.
 */

type Entry = { value: unknown; expiresAt: number };

const store = new Map<string, Entry>();
const inflight = new Map<string, Promise<unknown>>();

const DEFAULT_TTL_MS = 5 * 60 * 1000;

export function cacheGet<T>(key: string): T | undefined {
  const entry = store.get(key);
  if (!entry) return undefined;
  if (entry.expiresAt <= Date.now()) {
    store.delete(key);
    return undefined;
  }
  return entry.value as T;
}

export function cacheSet<T>(key: string, value: T, ttlMs: number = DEFAULT_TTL_MS): void {
  store.set(key, { value, expiresAt: Date.now() + ttlMs });
}

export function cacheDelete(key: string): void {
  store.delete(key);
}

export function cacheInvalidatePrefix(prefix: string): void {
  for (const key of Array.from(store.keys())) {
    if (key.startsWith(prefix)) store.delete(key);
  }
}

export function cacheClear(): void {
  store.clear();
  inflight.clear();
}

export async function cachedCall<T>(
  key: string,
  fetcher: () => Promise<T>,
  ttlMs: number = DEFAULT_TTL_MS,
): Promise<T> {
  const hit = cacheGet<T>(key);
  if (hit !== undefined) return hit;

  const pending = inflight.get(key) as Promise<T> | undefined;
  if (pending) return pending;

  const promise = (async () => {
    try {
      const value = await fetcher();
      cacheSet(key, value, ttlMs);
      return value;
    } finally {
      inflight.delete(key);
    }
  })();
  inflight.set(key, promise);
  return promise;
}

export function cacheKey(...parts: Array<string | number | null | undefined | Record<string, unknown>>): string {
  return parts
    .map((p) => {
      if (p == null) return '';
      if (typeof p === 'object') return JSON.stringify(p);
      return String(p);
    })
    .join(':');
}

File: agroamigo-iphone/lib/services/cache.dart
import 'dart:async';

/// In-memory TTL cache with in-flight de-duplication. Port of
/// `agroamigo-app/src/lib/cache.ts`.
class _Entry {
  final Object? value;
  final int expiresAtMs;
  _Entry(this.value, this.expiresAtMs);
}

class AppCache {
  AppCache._();
  static final AppCache instance = AppCache._();

  static const int defaultTtlMs = 5 * 60 * 1000;

  final Map<String, _Entry> _store = {};
  final Map<String, Future<Object?>> _inflight = {};

  T? get<T>(String key) {
    final entry = _store[key];
    if (entry == null) return null;
    if (entry.expiresAtMs <= DateTime.now().millisecondsSinceEpoch) {
      _store.remove(key);
      return null;
    }
    return entry.value as T?;
  }

  void set<T>(String key, T value, {int ttlMs = defaultTtlMs}) {
    _store[key] = _Entry(value, DateTime.now().millisecondsSinceEpoch + ttlMs);
  }

  void delete(String key) => _store.remove(key);

  void invalidatePrefix(String prefix) {
    _store.removeWhere((k, _) => k.startsWith(prefix));
  }

  void clear() {
    _store.clear();
    _inflight.clear();
  }

  Future<T> cachedCall<T>(
    String key,
    Future<T> Function() fetcher, {
    int ttlMs = defaultTtlMs,
  }) async {
    final hit = get<T>(key);
    if (hit != null) return hit;

    final pending = _inflight[key];
    if (pending != null) return await pending as T;

    final future = (() async {
      try {
        final value = await fetcher();
        set<T>(key, value, ttlMs: ttlMs);
        return value as Object?;
      } finally {
        _inflight.remove(key);
      }
    })();
    _inflight[key] = future;
    return (await future) as T;
  }
}

String cacheKey(List<Object?> parts) {
  return parts.map((p) {
    if (p == null) return '';
    return p.toString();
  }).join(':');
}
